thorlcr/activities/indexread/thindexread.cpp

```diff
@@ -313,46 +324,34 @@ class CIndexReadBase : public CMasterActivity
 
 class CIndexReadActivityMaster : public CIndexReadBase
 {
+    typedef CIndexReadBase PARENT;
+
     IHThorIndexReadArg *helper;
 
     void processKeyedLimit()
     {
-        rowcount_t total = aggregateToLimit();
-        if (total > limit)
+        if (PARENT::processKeyedLimit())
         {
             if (0 == (TIRkeyedlimitskips & helper->getFlags()))
             {
                 if (0 == (TIRkeyedlimitcreates & helper->getFlags()))
                     helper->onKeyedLimitExceeded(); // should throw exception
             }
         }
-        CMessageBuffer msg;
-        msg.append(total);
-        ICommunicator &comm = queryJobChannel().queryJobComm();
-        unsigned slaves = container.queryJob().querySlaves();
-        unsigned s=0;
-        for (; s<slaves; s++)
-        {
-            verifyex(comm.send(msg, s+1, mpTag));
-        }
     }
 public:
     CIndexReadActivityMaster(CMasterGraphElement *info) : CIndexReadBase(info)
     {
         helper = (IHThorIndexReadArg *)queryHelper();
     }
-    virtual void init() override
-    {
-        CIndexReadBase::init();
-        if (!container.queryLocalOrGrouped())
-        {
-            if (helper->canMatchAny())
-                limit = (rowcount_t)helper->getKeyedLimit();
-        }
-    }
-    virtual void process()
+    virtual void process() override
     {
-        if (limit != RCMAX)
+        if (container.queryLocalOrGrouped())
+            return;
+        if (!helper->canMatchAny())
+            return;
+        keyedLimit = (rowcount_t)helper->getKeyedLimit();
```

Comment:
would be slightly cleaner if this assignment and condition was inside processKeyedLimit() - the same code occurs 3 times.

Reply:
getKeyedLimit() is not part of the base helper, so can't call it inside processKeyedLimit()